// vuepress-theme-hope/lib/client/modules/blog/components/ArticleItem.js
import { withBase } from "@vuepress/client";
import { defineComponent, h, toRef } from "vue";
import { RouterLink } from "vue-router";
import PageInfo from "@theme-hope/modules/info/components/PageInfo";
import { SlideIcon, StickyIcon, } from "@theme-hope/modules/blog/components/icons/index";
import { useArticleInfo } from "@theme-hope/modules/blog/composables/index";
import { LockIcon } from "@theme-hope/modules/encrypt/components/icons";
import "../styles/article-item.scss";
export default defineComponent({
    name: "ArticleItem",
    props: {
        /**
         * Article information
         *
         * 文章信息
         */
        info: {
            type: Object,
            required: true,
        },
        /**
         * Article path
         *
         * 文章路径
         */
        path: { type: String, required: true },
    },
    setup(props) {
        const info = toRef(props, "info");
        const { info: articleInfo, items } = useArticleInfo(props);
        return () => h("div", { class: "article-item" }, h("article", {
            class: "article",
            vocab: "https://schema.org/",
            typeof: "Article",
        }, [
            info.value["u" /* ArticleInfoType.sticky */] ? h(StickyIcon) : null,
            h(RouterLink, { to: props.path }, () => [
                h("header", { class: "title" }, [
                    info.value["n" /* ArticleInfoType.isEncrypted */] ? h(LockIcon) : null,
                    info.value["y" /* ArticleInfoType.type */] === "s" /* PageType.slide */
                        ? h(SlideIcon)
                        : null,
                    h("span", { property: "headline" }, info.value.title),
                    info.value["v" /* ArticleInfoType.cover */]
                        ? h("meta", {
                            property: "image",
                            content: withBase(info.value["v" /* ArticleInfoType.cover */]),
                        })
                        : null,
                ]),
            ]),
            info.value["e" /* ArticleInfoType.excerpt */]
                ? h("div", {
                    class: "article-excerpt",
                    innerHTML: info.value["e" /* ArticleInfoType.excerpt */],
                })
                : null,
            h("hr", { class: "hr" }),
            h(PageInfo, {
                info: articleInfo.value,
                ...(items.value ? { items: items.value } : {}),
            }),
        ]));
    },
});
