// vuepress-theme-hope/lib/node/plugins/blog/category.js
import { defaultPageSorter } from "./utils.js";
export const getCategoryCategory = (options, themeData) => ({
    key: "category",
    getter: ({ routeMeta }) => routeMeta["c" /* ArticleInfoType.category */] || [],
    sorter: defaultPageSorter,
    path: options.category,
    layout: "BlogCategory",
    frontmatter: (localePath) => ({
        title: themeData.locales[localePath].blogLocales.category,
    }),
    itemPath: options.categoryItem,
    itemFrontmatter: (name, localePath) => ({
        title: `${name} ${themeData.locales[localePath].blogLocales.category}`,
    }),
    itemLayout: "BlogCategory",
});
export const getTagCategory = (options, themeData) => ({
    key: "tag",
    getter: ({ routeMeta }) => routeMeta["g" /* ArticleInfoType.tag */] || [],
    sorter: defaultPageSorter,
    path: options.tag,
    layout: "BlogCategory",
    frontmatter: (localePath) => ({
        title: themeData.locales[localePath].blogLocales.tag,
    }),
    itemPath: options.tagItem,
    itemLayout: "BlogCategory",
    itemFrontmatter: (name, localePath) => ({
        title: `${name} ${themeData.locales[localePath].blogLocales.tag}`,
    }),
});
