// vuepress-theme-hope/lib/client/components/SkipLink.js
import { defineComponent, h, ref, watch } from "vue";
import { useRoute } from "vue-router";
import { useThemeLocaleData } from "@theme-hope/composables/index";
import "../styles/skip-link.scss";
export default defineComponent({
    name: "SkipLink",
    props: {
        /** @description Content ID */
        content: {
            type: String,
            default: "main-content",
        },
    },
    setup(props) {
        const route = useRoute();
        const themeLocale = useThemeLocaleData();
        const backToTop = ref();
        watch(() => route.path, () => backToTop.value.focus());
        const focusMainContent = ({ target }) => {
            const el = document.querySelector(target.hash);
            if (el) {
                const removeTabIndex = () => {
                    el.removeAttribute("tabindex");
                    el.removeEventListener("blur", removeTabIndex);
                };
                el.setAttribute("tabindex", "-1");
                el.addEventListener("blur", removeTabIndex);
                el.focus();
                window.scrollTo(0, 0);
            }
        };
        return () => [
            h("span", {
                ref: backToTop,
                tabindex: "-1",
            }),
            h("a", {
                href: `#${props.content}`,
                class: "skip-link sr-only",
                onClick: focusMainContent,
            }, themeLocale.value.routeLocales.skipToContent),
        ];
    },
});
