// vuepress-theme-hope/lib/client/modules/outlook/components/ThemeColorPicker.js
import { defineComponent, h, onMounted } from "vue";
import "../styles/theme-color-picker.scss";
export default defineComponent({
    name: "ThemeColorPicker",
    props: {
        /**
         * Theme color picker config
         *
         * 主题色选择器配置
         */
        themeColor: {
            type: Object,
            required: true,
        },
    },
    setup(props) {
        const setThemeColor = (theme = "") => {
            const classes = document.documentElement.classList;
            const themes = Object.keys(props.themeColor).map((color) => `theme-${color}`);
            if (!theme) {
                localStorage.removeItem("theme");
                classes.remove(...themes);
                return;
            }
            classes.remove(...themes.filter((themeColorClass) => themeColorClass !== `theme-${theme}`));
            classes.add(`theme-${theme}`);
            localStorage.setItem("theme", theme);
        };
        onMounted(() => {
            const theme = localStorage.getItem("theme");
            if (theme)
                setThemeColor(theme);
        });
        return () => h("ul", { id: "theme-color-picker" }, [
            h("li", h("span", {
                class: "theme-color",
                onClick: () => setThemeColor(),
            })),
            ...Object.entries(props.themeColor).map(([color, value]) => h("li", h("span", {
                style: { background: value },
                onClick: () => setThemeColor(color),
            }))),
        ]);
    },
});
