// vuepress-theme-hope/lib/client/modules/info/components/CategoryInfo.js
import { defineComponent, h } from "vue";
import { useRoute, useRouter } from "vue-router";
import { generateIndexFromHash } from "vuepress-shared/client";
import { CategoryIcon } from "@theme-hope/modules/info/components/icons";
import { useMetaLocale } from "@theme-hope/modules/info/composables/index";
import "../styles/category.scss";
export default defineComponent({
    name: "CategoryInfo",
    inheritAttrs: false,
    props: {
        /**
         * Category information
         *
         * åˆ†ç±»ä¿¡æ¯
         */
        category: {
            type: Array,
            required: true,
        },
        /**
         * Whether in pure mode
         *
         * æ˜¯å¦å¤„äºŽçº¯å‡€æ¨¡å¼
         */
        pure: Boolean,
    },
    setup(props) {
        const router = useRouter();
        const route = useRoute();
        const metaLocale = useMetaLocale();
        const navigate = (event, path = "") => {
            if (path && route.path !== path) {
                event.preventDefault();
                void router.push(path);
            }
        };
        return () => props.category.length
            ? h("span", {
                class: "category-info",
                "aria-label": `${metaLocale.value.category}${props.pure ? "" : "ðŸŒˆ"}`,
                ...(props.pure ? {} : { "data-balloon-pos": "down" }),
            }, [
                h(CategoryIcon),
                ...props.category.map(({ name, path }) => h("span", {
                    class: [
                        "category-item",
                        {
                            // TODO: magic number 9 is tricky here
                            [`category${generateIndexFromHash(name, 9)}`]: !props.pure,
                            clickable: path,
                        },
                    ],
                    role: path ? "navigation" : "",
                    onClick: (event) => navigate(event, path),
                }, name)),
                h("meta", {
                    property: "articleSection",
                    content: props.category.map(({ name }) => name).join(","),
                }),
            ])
            : null;
    },
});
