// vuepress-theme-hope/lib/client/modules/info/components/DateInfo.js
import { ClientOnly, usePageLang } from "@vuepress/client";
import { defineComponent, h } from "vue";
import { CalendarIcon } from "@theme-hope/modules/info/components/icons";
import { useMetaLocale } from "@theme-hope/modules/info/composables/index";
export default defineComponent({
    name: "DateInfo",
    inheritAttrs: false,
    props: {
        /**
         * Date information
         *
         * æ—¥æœŸä¿¡æ¯
         */
        date: {
            type: Object,
            default: null,
        },
        /**
         * Localized date text
         *
         * æœ¬åœ°åŒ–çš„æ—¥æœŸæ–‡å­—
         */
        localizedDate: {
            type: String,
            default: "",
        },
        /**
         * Whether in pure mode
         *
         * æ˜¯å¦å¤„äºŽçº¯å‡€æ¨¡å¼
         */
        pure: Boolean,
    },
    setup(props) {
        const lang = usePageLang();
        const metaLocale = useMetaLocale();
        return () => props.date
            ? h("span", {
                class: "date-info",
                "aria-label": `${metaLocale.value.date}${props.pure ? "" : "ðŸ“…"}`,
                ...(props.pure ? {} : { "data-balloon-pos": "down" }),
            }, [
                h(CalendarIcon),
                h("span", h(ClientOnly, () => props.localizedDate ||
                    props.date.value?.toLocaleDateString(lang.value))),
                h("meta", {
                    property: "datePublished",
                    // ISO Format Date string
                    content: props.date?.value?.toISOString() || "",
                }),
            ])
            : null;
    },
});
