// vuepress-theme-hope/lib/client/modules/info/components/TagInfo.js
import { defineComponent, h } from "vue";
import { useRoute, useRouter } from "vue-router";
import { generateIndexFromHash } from "vuepress-shared/client";
import { TagIcon } from "@theme-hope/modules/info/components/icons";
import { useMetaLocale } from "@theme-hope/modules/info/composables/index";
import "../styles/tag.scss";
export default defineComponent({
    name: "TagInfo",
    inheritAttrs: false,
    props: {
        /**
         * Tag information
         *
         * æ ‡ç­¾ä¿¡æ¯
         */
        tag: {
            type: Array,
            default: () => [],
        },
        /**
         * Whether in pure mode
         *
         * æ˜¯å¦å¤„äºŽçº¯å‡€æ¨¡å¼
         */
        pure: Boolean,
    },
    setup(props) {
        const router = useRouter();
        const route = useRoute();
        const metaLocale = useMetaLocale();
        const navigate = (event, path = "") => {
            if (path && route.path !== path) {
                event.preventDefault();
                void router.push(path);
            }
        };
        return () => props.tag.length
            ? h("span", {
                class: "tag-info",
                "aria-label": `${metaLocale.value.tag}${props.pure ? "" : "ðŸ·"}`,
                ...(props.pure ? {} : { "data-balloon-pos": "down" }),
            }, [
                h(TagIcon),
                ...props.tag.map(({ name, path }) => h("span", {
                    class: [
                        "tag-item",
                        {
                            // TODO: magic number 9 is tricky here
                            [`tag-item${generateIndexFromHash(name, 9)}`]: !props.pure,
                            clickable: path,
                        },
                    ],
                    role: path ? "navigation" : "",
                    onClick: (event) => navigate(event, path),
                }, name)),
                h("meta", {
                    property: "keywords",
                    content: props.tag.map(({ name }) => name).join(","),
                }),
            ])
            : null;
    },
});
