// vuepress-theme-hope/lib/client/modules/info/components/WordInfo.js
import { computed, defineComponent, h } from "vue";
import { useLocaleConfig } from "vuepress-shared/client";
import { WordIcon } from "@theme-hope/modules/info/components/icons";
import { useMetaLocale } from "@theme-hope/modules/info/composables/index";
import { readingTimeLocales } from "@theme-hope/modules/info/utils/index";
export default defineComponent({
    name: "ReadTimeInfo",
    inheritAttrs: false,
    props: {
        /**
         * Reading time information
         *
         * é˜…è¯»æ—¶é—´ä¿¡æ¯
         */
        readingTime: {
            type: Object,
            default: () => null,
        },
        /**
         * Whether in pure mode
         *
         * æ˜¯å¦å¤„äºŽçº¯å‡€æ¨¡å¼
         */
        pure: Boolean,
    },
    setup(props) {
        const metaLocale = useMetaLocale();
        const readingTimeLocale = useLocaleConfig(readingTimeLocales);
        const words = computed(() => props.readingTime?.words.toString());
        const wordText = computed(() => readingTimeLocale.value.word.replace("$word", words.value || ""));
        return () => words.value
            ? h("span", {
                class: "words-info",
                "aria-label": `${metaLocale.value.words}${props.pure ? "" : "ðŸ” "}`,
                ...(props.pure ? {} : { "data-balloon-pos": "down" }),
            }, [
                h(WordIcon),
                h("span", wordText.value),
                h("meta", {
                    property: "wordCount",
                    content: words.value,
                }),
            ])
            : null;
    },
});
