// vuepress-theme-hope/lib/node/plugins/pageConverter.js
import { injectLocalizedDate } from "vuepress-shared";
import { convertFrontmatter } from "../compact/index.js";
import { checkFrontmatter } from "../frontmatter/check.js";
export const injectPageInfo = (page) => {
    const { filePathRelative } = page;
    const frontmatter = page.frontmatter;
    const isArticle = 
    // declaring this is an article
    frontmatter.article ||
        // generated from markdown files
        Boolean(frontmatter.article !== false && filePathRelative);
    const isSlide = frontmatter.layout === "Slide";
    // save page type to routeMeta
    page.routeMeta["y" /* ArticleInfoType.type */] = frontmatter.home
        ? "h" /* PageType.home */
        : isSlide
            ? "s" /* PageType.slide */
            : isArticle
                ? "a" /* PageType.article */
                : "p" /* PageType.page */;
    // save relative file path into page data to generate edit link
    page.data.filePathRelative = filePathRelative;
    page.routeMeta["title"] = page.title;
    if ("icon" in frontmatter)
        page.routeMeta["i" /* ArticleInfoType.icon */] = frontmatter.icon;
    // resolve shortTitle
    if ("shortTitle" in frontmatter)
        page.routeMeta["s" /* ArticleInfoType.shortTitle */] = frontmatter.shortTitle;
};
export const extendsPagePlugin = (legacy = false) => ({
    name: "vuepress-theme-hope-extends-page",
    extendsPage: (page, app) => {
        if (legacy)
            page.frontmatter = convertFrontmatter(page.frontmatter, page.filePathRelative || "");
        checkFrontmatter(page, app.env.isDebug);
        injectPageInfo(page);
        injectLocalizedDate(page);
    },
});
export const useExtendsPagePlugin = (app, legacy = false) => {
    app.use(extendsPagePlugin(legacy));
};
